conjure-undertow-annotations/src/main/java/com/palantir/conjure/java/undertow/annotations/CookieDeserializer.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * (c) Copyright 2022 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.conjure.java.undertow.annotations;
+
+import com.palantir.conjure.java.undertow.lib.Deserializer;
+import com.palantir.logsafe.Preconditions;
+import io.undertow.server.HttpServerExchange;
+import io.undertow.server.handlers.Cookie;
+import java.io.IOException;
+
+public final class CookieDeserializer<T> implements Deserializer<T> {
+
+    private final String cookieName;
+    private final ParamDecoder<? extends T> decoder;
+
+    public CookieDeserializer(String cookieName, ParamDecoder<? extends T> decoder) {
+        this.cookieName = Preconditions.checkNotNull(cookieName, "Cookie name is required");
+        this.decoder = Preconditions.checkNotNull(decoder, "Decoder is required");
+    }
+
+    @Override
+    public T deserialize(HttpServerExchange exchange) throws IOException {
+        Cookie cookie = exchange.getRequestCookie(cookieName);
+        if (cookie == null) {
+            // Let the decoder handle the null value (e.g. fail or return 'Optional#empty').
+            return decoder.decode(null);
+        }
+        return decoder.decode(cookie.getValue());
```

Comment:
We try not to allow `null` in our APIs (direct interaction with the undertow apis is a bit different though). What do you think about throwing instead?

decoder taking an Optional would make it a bit harder to interact with conjure-undertow plainserde, but we could add a default `T handleMissing()` which throws.

Reply:
Hmm yeah, I forwarded the `null` here to support optional cookie params (i.e. `@Cookie(...) Optional<String>`).

I tried adding a separate Serializer for optional cookies so we can handle the `null` case explicitly. That way we don't really have to touch the decoders. But it's a bit of duplication between the `CookieDeserializer` and `OptionalCookieDeserializer`. Lmk what you think!

Reply:
Hmm, the additional types+conditionals are a bit odd, I think I'd prefer handling all cookies using the same `CookieDeserializer`. That way we can also support types like `OptionalInt` and alias types around an optional values.

What do you think about adding something like this to `ParamDecoder`?
```java
/** May return a value which should be used if the request doesn't include the requested parameter. */
default Optional<T> noValuePresent() {
    return Optional.empty();
}

Reply:
To clarify the snippet, returning an empty optional would allow the deserializer calling into the decoder to throw the same exception it does by default.

Reply:
I think that makes sense, didn't think of the `OptionalInt` case before. Pushed up a new change that adds the `noValuePresent` default method to the `ParamDecoder`. For all "optional" param decoders, we would return an `Optional.empty()` here and for all "non-optional" param decoders, we would throw. But happy to iterate!